### DISCUSSION TITLE
Best Way to Manage Multiple Related Forms at Multiple Hierarchical Levels

### DISCUSSION THREAD

**Discussion by GeekLad** [Budibase/budibase#10889] [Help]
I have many entities with related fields within various levels of a hierarchy.  When creating a new record, I want to be able to either select the related record or create it.  Below the autocomplete related field, I add a button to create a new related record.  

In an ideal world, when that button is clicked, a modal could be used to create the new related record and when they save it, the modal is closed and the related field is then populated with the new related record.

However, I know that unfortunately the app doesn't function in that way, because the related fields are populated when the form initially loads and they cannot be updated when the modal closes.  I've seen others request such functionality, and I really hope that this eventually gets implemented.

It would be a much better user experience to use a modal, rather than having to navigate to a new form, save that form, and then navigate back to the original form, reloading it in the process.  It's not an ideal user experience and results in more load on the server than should be necessary, but I'm dealing with it.

I'm hoping that perhaps someone can provide some tips on a better way to manage such a use case.  Let's say I have a contact record, and I want to create a related location.  In that scenario, here's what I'm currently doing:

- The default for the location field is set to a **NewLocation** state variable
- On each field in the contact form, on the update action I set the state for that field
- On each field in the contact form, I set the default value to be each of the associated states
- I execute the following actions when the user clicks the **Create New Location** button on the new contact form
  - Set a **LocationNavigation** state variable to remember the path to return back to (which in this scenario is **/contacts/new**)
  - Navigate to **/locations/new**
- On save, I set the **NewLocation** state variable
- I navigate back to the return path state variable

Now this sounds simple enough, but when there are multiple levels of relations, it gets to be quite complex and requires a lot of state management.  In my app, I have a **Project** table with a **Project Location** related field, a **Billing Location** related field (both of these locations reside in a single **Location** table), a **Project Contact**, **Property Owner** (both residing in the **Contacts** table), etc...  Each of the contacts has a related location, and the project also has a related location.  The hierarchy is something like this:

- Project
  - Project Location (Location Table)
  - Billing Location (Location Table)
  - Project Contact (Contact Table)
    - Location (Location Table)
  - Property Owner (Contact Table)
    - Location (Location Table)

It involves keeping track of all the fields in state variables, deleting the states these when I want to add new related records, keeping track of all the fields/states, keeping track of states for the original related record, etc...  I feel as though there must be a better way to manage such a use case, as managing all of these states is becoming quite cumbersome.

I was thinking of perhaps storing everything in a single state variable using `JSON.stringify()`/`JSON.parse()` to reduce the number of **Update State** actions I need to run every time I need to add/update records.  I could write some reusable JS functions, and just copy/paste them into the Update State actions everywhere necessary, and adjust the inputs.  It's a lot of copy/paste/anti-DRY, but it may be slightly better than having to add so many repetitive **Update State** actions and manage so many different state variables.

Does anyone have any suggestions for a better way of managing such a use case?

**Comment by melohagan** [comment]
Hey @GeekLad 

If you use a **custom** [form schema](https://docs.budibase.com/docs/forms#form-schema) then you can save to multiple linked tables in one form:

 - https://docs.budibase.com/docs/chaining-actions#example-creating-and-linking-two-table-entries-in-one-form
 - https://docs.budibase.com/docs/form-steps

**Reply by GeekLad** [reply]
That makes sense, and that would work if I wanted to create new related records every time I created a new project (create a new location, and then attach to the new project, using multiple save row steps in the save action).  However, I would like to have the option to either create new related records or select related existing records.

**Reply by melohagan** [reply]
You could possibly have a checkbox that toggles state in your form. If checked then a relationship picker is shown that allows you to select existing records. If unchecked, then display a button that opens a modal for creating a new related entry.
https://docs.budibase.com/docs/conditions
